### PR TITLE
Feature/dcc 5684 kv search

### PR DESCRIPTION
**Pull request by rtisma**
- info search client which uses the same subcommand as for idsearch (called search), however has logic to deduce which search mode to use (IdSearch or InfoSearch modes) and detects when 2 mutually exclusive switches are defined and writes to stderr
- added search term syntax error to lists of server errors for song-server
- added info search enpoints, as well as infoSearch service call
- added SearchQueryBuilder which builds the searchQuery string for the searchRepository
- added searchRepository for searching
- added feature to enable/disable "info" field in InfoSearchResponse. The reasoning is that the info field may make the payload too big.

### CODE REVIEW

**Comment by rtisma** [comment]
added this for building purposes. will notify when ready. also adding unit tests now

**Comment by rtisma** [comment]
ok to test

**Comment by rtisma** [comment]
test this please

**Comment by andricDu** [comment]
test this please

**Comment by andricDu** [comment]
test this please

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    assertThat(actualResponseList).hasSize(1);
+    val infoSearchResponse = actualResponseList.get(0);
+    assertThat(this.analysisRespMap).containsKey(infoSearchResponse.getAnalysisId());
+    assertThat(infoSearchResponse).isEqualTo(this.analysisRespMap.get(infoSearchResponse.getAnalysisId()));
+  }
+
+  @Ignore
+  @Test
+  public void testGreedy(){
+    // 1) key1 = miss and key1 = a$
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete this whole test

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    //    - data1 has key1 = mississauga, key2 = base
+    //    - data2 has key1 = missippi, key2 = missa
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testNonGreedy(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    // test "^rob$" and "^ro$"
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testNested(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    //        "address" : { "coordinate" : { "latitude" : 17.0 ,  "longitude" : 32.0 } }
+    //      }
+    // 1)  key1 is address->coordinate->latitude
+    // 2)  key1 is name
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testANDing(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    //    - data2 has key1 = missippi, key2 = male
+    //    - data3 has key1 = missippi, key2 = unknown
+    //    - data3 has key1 = toronto, key2 = male
+    //    - data3 has key1 = toronto, key2 = unknown
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testIncludeInfo(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    // test that response has correct fields depending on includeInfo value
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testExcludeInfo(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    // test that response has correct fields depending on includeInfo value
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testSyntaxErrors(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    // create malformed search term if possible
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testArrayNonGreedy(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

File: song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    // test that a field that contains an array, can be searched for a non-greedy pattern, and if one or more items
+    // it, the analysis id is returned. for example. info.info->>'study' ~ '^PCA$'
+    fail("not implemented yet");
+  }
+
+  @Ignore
+  @Test
+  public void testArrayGreedy(){
```
Comment (andricDu) [inline song-server/src/test/java/org/icgc/dcc/song/server/repository/search/InfoSearchTest.java]: delete

**Comment by andricDu** [comment]
test this please

**Comment by rtisma** [comment]
test this please

**Comment by rtisma** [comment]
@andricDu i ignore the 4 codacy 4 errors. it says the functions dont assert or fail. Those functions call a function called `runTest` which has an assertThat statement, hence why i ignored the errors

**Pull request merged**